dao/src/main/java/com/epam/esm/dao/GiftCertificateDao.java

```diff
@@ -0,0 +1,101 @@
+package com.epam.esm.dao;
+
+import com.epam.esm.entity.GiftCertificate;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * This interface represents an api to interact with the GiftCertificate entity in database.
+ * Implementations : {@link com.epam.esm.dao.impl.GiftCertificateDaoImpl} classes.
+ *
+ * @since 1.0
+ */
+public interface GiftCertificateDao {
+
+    /**
+     * This method saves GiftCertificate entity.
+     *
+     * @param certificate GiftCertificate entity without id.
+     * @return GiftCertificate entity with generated id.
+     * @since 1.0
+     */
+    GiftCertificate save(GiftCertificate certificate);
+
+
+    /**
+     * This method get GiftCertificate entity by id.
+     *
+     * @param id GiftCertificate entity's id.
+     * @return Optional of GiftCertificate entity.
+     * @since 1.0
+     */
+    Optional<GiftCertificate> getById(long id);
+
+    /**
+     * This method get all GiftCertificate entities.
+     *
+     * @return List of all GiftCertificate entities.
+     * @since 1.0
+     */
+    List<GiftCertificate> findAll();
+
+
+    /**
+     * This method get GiftCertificate entities by part of their description.
+     *
+     * @param partOfDescription part of GiftCertificate entity's description for search
+     * @return List of GiftCertificate entities which was found by part of description.
+     * @since 1.0
+     */
+    List<GiftCertificate> getAllByPartOfDescription(String partOfDescription);
```

Comment:
I suppose it may be improved. You created a lot of get...() methods, but what if you have findBy(Seacrh criteria)?
By this approach you have only one method for searching and you can create any new searching criteria in the future and use this method

Reply:
Remove all getAll*() methods. Replaced it by Criterias. Servica layer requests concrete Criteria with concrete Parameters(fields for searching) from CriteriaFactory and passes it to Dao layer. Sorting now is a part of Service layer and uses Comparators instead of ORDER BY sql queries.

commit cb8a000